File: crates/api/src/dto/library/unmatched_file.rs
use serde::Serialize;

use domain::library::{MatchCandidate, UnmatchedFile};

use crate::dto::common::TitleRefDto;

#[derive(Debug, Serialize)]
pub struct MatchCandidateDto {
    pub title: TitleRefDto,
    pub confidence: f32,
    pub label: String,
}

impl From<MatchCandidate> for MatchCandidateDto {
    fn from(c: MatchCandidate) -> Self {
        MatchCandidateDto {
            title: c.title.into(),
            confidence: c.confidence,
            label: c.label,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct UnmatchedFileResponse {
    pub id: String,
    pub library_id: String,
    pub path: String,
    pub candidates: Vec<MatchCandidateDto>,
    pub created_at: String,
    pub updated_at: String,
}

impl From<UnmatchedFile> for UnmatchedFileResponse {
    fn from(f: UnmatchedFile) -> Self {
        UnmatchedFileResponse {
            id: f.id.0,
            library_id: f.library.0,
            path: f.path,
            candidates: f.candidates.into_iter().map(Into::into).collect(),
            created_at: f.created_at.to_string(),
            updated_at: f.updated_at.to_string(),
        }
    }
}

File: clients/flutter/lib/model/library/unmatched_file.dart
import 'package:freezed_annotation/freezed_annotation.dart';

import 'package:shadowmask/model/common/title_ref.dart';

part 'unmatched_file.freezed.dart';
part 'unmatched_file.g.dart';

@freezed
abstract class MatchCandidate with _$MatchCandidate {
  const factory MatchCandidate({
    required TitleRef title,
    @Default(0) double confidence,
    @Default('') String label,
  }) = _MatchCandidate;

  factory MatchCandidate.fromJson(Map<String, dynamic> json) =>
      _$MatchCandidateFromJson(json);
}

@freezed
abstract class UnmatchedFile with _$UnmatchedFile {
  const factory UnmatchedFile({
    required String id,
    required String libraryId,
    required String path,
    @Default(<MatchCandidate>[]) List<MatchCandidate> candidates,
    required String createdAt,
    required String updatedAt,
  }) = _UnmatchedFile;

  factory UnmatchedFile.fromJson(Map<String, dynamic> json) =>
      _$UnmatchedFileFromJson(json);
}
